src/commonMain/kotlin/org/treeWare/model/operator/Difference.kt

```diff
@@ -0,0 +1,407 @@
+
+    /*If a result tree is empty, then instead make the value of the tree null*/
+    val addTree = when{
+        differenceVisitor.stepsFromAdd > 0 -> null
+        else -> differenceVisitor.addMainModel
+    }
+    val deleteTree = when{
+        differenceVisitor.stepsFromDelete > 0 -> null
+        else -> differenceVisitor.deleteMainModel
+    }
+    return listOf(addTree, deleteTree)
+}
+
+private class DifferenceVisitor : AbstractLeaderManyModelVisitor<TraversalAction>(
+    TraversalAction.CONTINUE
+) {
+    var stepsFromAdd = 0 //the number of steps away from the path from the root to an added node
+    var stepsFromDelete = 0 //the number of steps away from the path from the root to a deleted node
+    var stepsFromUpdate = 0 //the number of steps away from the path from the root to an updated node
+    val addStack = ArrayDeque<MutableElementModel>()
+    val deleteStack = ArrayDeque<MutableElementModel>()
+    val updateStack = ArrayDeque<MutableElementModel>()
+    lateinit var addMainModel: MutableMainModel
+    lateinit var deleteMainModel: MutableMainModel
+    lateinit var updateMainModel: MutableMainModel
+
+    override fun visitMain(leaderMainList: List<MainModel?>): TraversalAction {
+        addMainModel = MutableMainModel(leaderMainList.last()?.mainMeta)//what needs to be added
+        deleteMainModel = MutableMainModel(leaderMainList.first()?.mainMeta)//what needs to be removed
+        updateMainModel = MutableMainModel(leaderMainList.last()?.mainMeta)//what needs to be updated
+
+        addStack.addFirst(addMainModel)
+        deleteStack.addFirst(deleteMainModel)
+        updateStack.addFirst(updateMainModel)
+        stepsFromAdd ++
+        stepsFromDelete ++
+        stepsFromUpdate ++
+
+        return TraversalAction.CONTINUE
+    }
+
+    override fun leaveMain(leaderMainList: List<MainModel?>) {
+        addStack.removeFirst()
+        assertInDevMode(addStack.isEmpty())
+        deleteStack.removeFirst()
+        assertInDevMode(deleteStack.isEmpty())
+        updateStack.removeFirst()
+        assertInDevMode(updateStack.isEmpty())
+    }
+
+    override fun visitEntity(leaderEntityList: List<EntityModel?>): TraversalAction {
+
+        /**===== Build new entities for the trees =====**/
+        val addParent = addStack.first()
+        val addEntity = newDisconnectedValue(addParent)
+
+        addStack.addFirst(addEntity)
+        stepsFromAdd ++
+
+        val deleteParent = deleteStack.first() //add the entity to the stack
+        val deleteEntity = newDisconnectedValue(deleteParent)
+        deleteStack.addFirst(deleteEntity)
+        stepsFromDelete ++
+
+        val updateParent = updateStack.first()
+        val updateEntity = newDisconnectedValue(updateParent)
+        updateStack.addFirst(updateEntity)
+        stepsFromUpdate ++
+
+        /**=====Check if it's absent in the original=====**/
+        if(leaderEntityList.first() == null) { //if the original doesn't have the entity:
```

Comment:
`copy()` operator can copy entire sub-trees. So when you find the root of an added/deleted sub-tree, you can copy over the entire sub-tree, and then return `TraversalAction.ABORT_SUB_TREE` to prevent `forEach()` from traversing the sub-tree. WARNING: the corresponding leave-method will be called even if you abort the sub-tree in a visit method.